polyfill/intersection-observer.js

```diff
@@ -0,0 +1,696 @@
+    var targetRect = getBoundingClientRect(target);
+    var rootContainsTarget = this._rootContainsTarget(target);
+    var oldEntry = item.entry;
+    var newEntry = item.entry = new IntersectionObserverEntry({
+      time: now(),
+      target: target,
+      boundingClientRect: targetRect,
+      rootBounds: rootRect,
+      intersectionRect: (rootIsInDom && rootContainsTarget) ?
+          this._computeTargetAndRootIntersection(target, rootRect) :
+          getEmptyRect()
+    });
+
+    if (rootIsInDom && rootContainsTarget) {
+      // If the new entry intersection ratio has crossed any of the
+      // thresholds, add a new entry.
+      if (this._hasCrossedThreshold(oldEntry, newEntry))  {
+        // Record that this target is in the DOM, a child or root, and
+        // is intersecting with the root.
+        item.hasIntersection = hasIntersection(newEntry.intersectionRect);
+        this._queuedEntries.push(newEntry);
+      }
+    } else {
+      // If the root is not in the DOM or target is not contained within
+      // root but the previous entry for this target had an intersection,
+      // add a new record indicating removal.
+      if (item.hasIntersection) {
+        item.hasIntersection = false;
+        this._queuedEntries.push(newEntry);
+      }
+    }
+  }.bind(this));
+
+  if (this._queuedEntries.length) {
+
+    this._callback(this.takeRecords(), this);
+  }
+};
+
+
+/**
+ * Accepts a target and root rect computes the intersection between then
+ * following the algorithm in the spec.
+ * TODO(philipwalton): at this time clip-path is not considered.
+ * https://wicg.github.io/IntersectionObserver/#calculate-intersection-rect-algo
+ * @param {Element} target The target DOM element
+ * @param {Object} rootRect The bounding rect of the root after being
+ *     expanded by the rootMargin value.
+ * @return {Object} The final intersection rect object.
+ * @private
+ */
+IntersectionObserver.prototype._computeTargetAndRootIntersection =
+    function(target, rootRect) {
+
+  var targetRect = getBoundingClientRect(target);
+  var intersectionRect = targetRect;
+  var parent = target.parentNode;
+  var notAtRoot = true;
+
+  while (notAtRoot) {
+    var parentRect = null;
+
+    // If we're at the root element, set parentRect to the already
+    // calculated rootRect.
+    if (parent == this.root || parent.nodeType != 1) {
+      notAtRoot = false;
+      parentRect = rootRect;
+    }
+    // Otherwise check to see if the parent element hides overflow,
+    // and if so update parentRect.
+    else {
+      var style = window.getComputedStyle(parent);
+      if (style.overflow != 'visible') {
+        parentRect = getBoundingClientRect(parent);
+      }
+    }
+    // If either of the above conditionals set a new parentRect,
+    // calculate new intersection data.
+    if (parentRect) {
+      intersectionRect = computeRectIntersection(
+          parentRect, intersectionRect);
+    }
+    parent = parent.parentNode;
+  }
```

Comment:
Early-out here if there is no intersection.  Using edge-inclusive intersection makes this more straightforward (see comment on computeRectIntersection below).